src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -413,16 +326,139 @@ public ValueTask<(Solution solution, bool updated)> TryUpdateWorkspaceCurrentSol
+            /// </summary>
+            private Checksum? _lastRequestedChecksum;
+            private ReferenceCountedDisposable<LazySolution>? _lastRequestedSolution;
+
+            /// <summary>
+            /// Mapping from solution-checksum to the solution computed for it.  This is used so that we can hold a
+            /// solution around as long as the checksum for it is being used in service of some feature operation (e.g.
+            /// classification).  As long as we're holding onto it, concurrent feature requests for the same solution
+            /// checksum can share the computation of that particular solution and avoid duplicated concurrent work.
+            /// </summary>
+            private readonly Dictionary<Checksum, ReferenceCountedDisposable<LazySolution>> _solutionChecksumToLazySolution = new();
+
+            public ChecksumToSolutionCache(SemaphoreSlim gate)
+            {
+                _gate = gate;
+            }
+
+            public async ValueTask<ReferenceCountedDisposable<LazySolution>?> TryFastGetSolutionAsync(
+                Checksum solutionChecksum,
+                CancellationToken cancellationToken)
+            {
+                Contract.ThrowIfNull(solutionChecksum);
+
+                using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+                {
+                    // Adding refcounts in the cases where we see we have a direct reference to the solution cannot
+                    // fail.  We only have a matching item in _lastRequestedChecksum if it had a positive refcount
+                    // already added to it when it was placed in this cache.  And the only thing that releases that
+                    // refcount item is code that runs while holding the same _gate that we are holding, and then
+                    // overwrites the item.  So it's not possible for us to race with anything on this, and so we must
+                    // succeed.
+
+                    if (_lastRequestedChecksum == solutionChecksum)
+                        return _lastRequestedSolution!.TryAddReference() ?? throw ExceptionUtilities.Unreachable;
+
+                    // Next, see if it's the map of items being looked at by another request.  Note: this may fail to add
+                    // the ref count as we might see an item that is disposed of, but hasn't been cleaned from this cache
+                    // yet.
+                    if (_solutionChecksumToLazySolution.TryGetValue(solutionChecksum, out var refCountedLazySolution))
+                        return refCountedLazySolution.TryAddReference();
+
+                    return null;
+                }
+            }
+
+            public async ValueTask<ReferenceCountedDisposable<LazySolution>> SlowGetOrCreateSolutionAsync(
+                Checksum solutionChecksum, Func<CancellationToken, Task<Solution>> getSolutionAsync, CancellationToken cancellationToken)
```

Comment:
this is the same as before, except that getSolutionAsync can be passed in since there is different behavior for the anyBranch cache versus the primaryBranchCache.